src/redis_rdb.cc

```diff
@@ -0,0 +1,429 @@
+    }
+    std::string ns, user_key;
+    ExtractNamespaceKey(iter->key(), &ns, &user_key, storage_->IsSlotIdEncoded());
+    int32_t ttl = metadata.TTL();
+    RedisType type = metadata.Type();
+
+    s = util.SaveMetaKeyPair(user_key, ttl, type);
+    if (!s.IsOK()) return s;
+    switch (type) {
+        case kRedisString:
+            rs = string_db.Get(user_key, &value);
+            if (rs.ok()) {
+               s = util.SaveStringObject(value);
+               if (!s.IsOK()) return s;
+            }
+            break;
+        case kRedisHash:
+            rs = hash_db.GetAll(user_key, &field_values, HashFetchType::kAll);
+            if (rs.ok()) {
+                s = util.SaveLen(field_values.size());
+                if (!s.IsOK()) return s;
+                for (const auto &fv : field_values) {
+                    s = util.SaveStringObject(fv.field);
+                    if (!s.IsOK()) return s;
+                    s = util.SaveStringObject(fv.value);
+                    if (!s.IsOK()) return s;
+                }
+            }
+            break;
+        case kRedisList:
+            rs = list_db.Range(user_key, 0, -1, &elems);
+            if (rs.ok()) {
+                s = util.SaveLen(elems.size());
+                if (!s.IsOK()) return s;
+                for (const auto &item : elems) {
+                    s = util.SaveStringObject(item);
+                    if (!s.IsOK()) return s;
+                }
+            }
+            break;
+        case kRedisSet:
+            rs = set_db.Members(user_key, &members);
+            if (rs.ok()) {
+                s = util.SaveLen(members.size());
+                if (!s.IsOK()) return s;
+                for (const auto &item : members) {
+                    s = util.SaveStringObject(item);
+                    if (!s.IsOK()) return s;
+                }
+            }
+            break;
+        case kRedisZSet:
+            rs = zset_db.Range(user_key, 0, -1, 0, &member_scores);
+            if (rs.ok()) {
+                s = util.SaveLen(member_scores.size());
+                for (const auto &item : member_scores) {
+                    s = util.SaveStringObject(item.member);
+                    if (!s.IsOK()) return s;
+                    std::string score = std::to_string(item.score);
+                    s = util.SaveStringObject(score);
+                    if (!s.IsOK()) return s;
+                }
+            }
+            break;
+            case kRedisBitmap:
+            rs = bitmap_db.GetString(user_key, UINT32_MAX, &value);
+            if (rs.ok()) {
+                s = util.SaveStringObject(value);
+                if (!s.IsOK()) return s;
+            }
+            break;
+        default:
+            break;
+    }
+    if (!rs.ok()) return Status(Status::NotOK, rs.ToString());
+  }
+  s = util.SaveTail();
+  return s;
+}
+
+RedisDatabase::Util::Util(const std::string &file_name) {
+    file_.open(file_name, std::ofstream::binary);
```

Comment:
What happens if a file fails to open? Will the user be informed?

Reply:
Thank you.Do you think it's OK to throw an exception?